File: client-clojure/src/sirqul_io_t_platform/specs/billable_entity.clj
(ns sirqul-io-t-platform.specs.billable-entity
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.contact-info :refer :all]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.retailer :refer :all]
            [sirqul-io-t-platform.specs.account :refer :all]
            [sirqul-io-t-platform.specs.category :refer :all]
            [sirqul-io-t-platform.specs.filter :refer :all]
            [sirqul-io-t-platform.specs.asset :refer :all]
            [sirqul-io-t-platform.specs.album :refer :all]
            [sirqul-io-t-platform.specs.subscription :refer :all]
            [sirqul-io-t-platform.specs.third-party-network :refer :all]
            )
  (:import (java.io File)))


(def billable-entity-data
  {
   (ds/opt :id) int?
   (ds/opt :active) boolean?
   (ds/opt :valid) boolean?
   (ds/opt :name) string?
   (ds/opt :contactInfo) contact-info-spec
   (ds/opt :responsible) account-spec
   (ds/opt :retailers) (s/coll-of retailer-spec)
   (ds/opt :managers) (s/coll-of account-spec)
   (ds/opt :categories) (s/coll-of category-spec)
   (ds/opt :filters) (s/coll-of filter-spec)
   (ds/opt :media) (s/coll-of asset-spec)
   (ds/opt :mediaAlbum) album-spec
   (ds/opt :subscription) subscription-spec
   (ds/opt :disbursementIntrospect) third-party-network-spec
   (ds/opt :authorizeNetTransactionKey) string?
   (ds/opt :authorizeNetApiKey) string?
   })

(def billable-entity-spec
  (ds/spec
    {:name ::billable-entity
     :spec billable-entity-data}))

File: client-scalaz/src/main/scala/org/openapitools/client/api/BillableEntity.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import BillableEntity._

case class BillableEntity (
  id: Option[Long],
active: Option[Boolean],
valid: Option[Boolean],
name: Option[String],
contactInfo: Option[ContactInfo],
responsible: Option[Account],
retailers: Option[List[Retailer]],
managers: Option[List[Account]],
categories: Option[List[Category]],
filters: Option[List[Filter]],
media: Option[List[Asset]],
mediaAlbum: Option[Album],
subscription: Option[Subscription],
disbursementIntrospect: Option[ThirdPartyNetwork],
authorizeNetTransactionKey: Option[String],
authorizeNetApiKey: Option[String])

object BillableEntity {
  import DateTimeCodecs._

  implicit val BillableEntityCodecJson: CodecJson[BillableEntity] = CodecJson.derive[BillableEntity]
  implicit val BillableEntityDecoder: EntityDecoder[BillableEntity] = jsonOf[BillableEntity]
  implicit val BillableEntityEncoder: EntityEncoder[BillableEntity] = jsonEncoderOf[BillableEntity]
}
